pkg/config/remote/service/service.go

```diff
@@ -1014,61 +1080,105 @@ func (s *CoreAgentService) ClientGetConfigs(_ context.Context, request *pbgo.Cli
 		return nil, err
 	}
 
-	neededFiles, err := filterNeededTargetFiles(matchedClientConfigs, request.CachedTargetFiles, directorTargets)
+	cachedTargetsMap, err := makeFileMetaMap(request.CachedTargetFiles)
 	if err != nil {
 		return nil, err
 	}
 
-	// TODO: Do not hold the mutex while calling getTargetFiles -- it may go to
-	// disk or network.
-	targetFiles, err := getTargetFiles(s.mu.uptane, neededFiles)
-	if err != nil {
-		return nil, err
+	// Compute the set of configs that are explicitly needed for subscriptions.
+	subscriptionConfigs := make(map[string]struct{})
+	for _, config := range matchedClientConfigs {
+		if _, ok := needCompleteProducts[productFromPath(config)]; ok {
+			subscriptionConfigs[config] = struct{}{}
+		}
+	}
+
+	// Compute the set of configs that are needed for the client.
+	responseConfigs := filtered(matchedClientConfigs, func(config string) bool {
+		return tufutil.FileMetaEqual(
+			cachedTargetsMap[config],
+			directorTargets[config].FileMeta,
+		) != nil
+	})
+
+	// Make neededForSubscriptions only contain the configs that are not needed
+	// for the client.
+	for _, config := range responseConfigs {
+		delete(subscriptionConfigs, config)
+	}
+
+	allConfigs := slices.AppendSeq(responseConfigs, maps.Keys(subscriptionConfigs))
+
+	// Sort the configs to ensure a consistent order (helps with tests).
+	slices.Sort(allConfigs)
+	var allFiles []*pbgo.File
+	if len(allConfigs) > 0 {
+		allFiles, err = getTargetFiles(s.mu.uptane, allConfigs)
+		if err != nil {
+			return nil, err
+		}
+		// Move the extra files needed just for subscriptions to the end.
+		slices.SortStableFunc(allFiles, func(a, b *pbgo.File) int {
+			return cmp.Compare(
+				boolToInt(contains(subscriptionConfigs, a.Path)),
+				boolToInt(contains(subscriptionConfigs, b.Path)),
+			)
+		})
+	}
+	responseFiles := allFiles[:len(responseConfigs)]
+
+	// Notify subscriptions with all files (including cached ones for first-time
+	// subscriptions).
+	if len(interestedSubs) > 0 {
+		s.mu.subscriptions.notify(
+			interestedSubs,
+			request.Client,
+			matchedClientConfigs,
+			responseFiles,
+			allFiles,
+		)
+	}
```

Comment:
This section is a bit hard to follow unless you dive through it line by line. I think the comments are too much a translation of the "how" of the code, in English, rather than trying to describe the why/intent. This forces the reader into having to read every line to extract the intent.

I do appreciate how it's trying to efficiently pull config files, rather than track everything separately which might result in pulling a config file twice. And efficiency sometimes has a tradeoff of clarity. But I think a higher level block comment can restore the clarity and allow readers to only dive into the nitty gritty if they care and give us the best of both worlds.

Reply:
Fair. I tried to structure a top-level comment more systematically in the last commit. See if it pleases.